File: dox_development/dox_development/report/item_card_quantity_report/item_card_quantity_report.py
# Copyright (c) 2023, rakesh and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import msgprint, _


def execute(filters=None):
    filters_report = {
        "company": filters.get("company"),
        "from_date": filters.get("from_date"),
        "to_date": filters.get("to_date"),
        "warehouse": filters.get("warehouse"),
        "item_code": filters.get("item_code"),
        "batch_no": filters.get("batch_no"),
        "variant_of": filters.get("variant_of"),
    }
    from frappe.desk.query_report import run

    res = run("Stock Ledger", filters=filters_report)
    data = res.get("result")
    (
        total_in_qty,
        total_qty_out,
        total_sales_invoice,
        total_purchase,
        total_stock_entry,
        reconcilation,
    ) = (0.0, 0.0, 0.0, 0.0, 0.0, 0.0)
    for row in data:
        if row.get("voucher_type") == "Sales Invoice":
            total_sales_invoice += 1
        elif (
            row.get("voucher_type") == "Purchase Receipt"
            or row.get("voucher_type") == "Purchase Invoice"
        ):
            total_purchase += 1
        elif row.get("voucher_type") == "Stock Entry":
            total_stock_entry += 1
        reconcilation = total_sales_invoice - total_purchase + total_stock_entry
        total_in_qty += row.get("in_qty", 0.0)
        total_qty_out += row.get("out_qty", 0.0)
        row["total_in"] = total_in_qty
        row["total_sales"] = total_sales_invoice
        row["total_purchase"] = total_purchase
        row["total_stock"] = total_stock_entry
        row["total_out"] = total_qty_out
        row["reconcilation"] = reconcilation

    columns = [
        {
            "label": _("Date"),
            "fieldname": "posting_date",
            "fieldtype": "Date",
            "width": 150,
        },
        {
            "label": _("UOM"),
            "fieldname": "stock_uom",
            "fieldtype": "Link",
            "options": "UOM",
            "width": 100,
        },
        {
            "label": _("In Qty"),
            "fieldname": "in_qty",
            "fieldtype": "Float",
            "width": 100,
            "convertible": "qty",
        },
        {
            "label": _("Out Qty"),
            "fieldname": "out_qty",
            "fieldtype": "Float",
            "width": 100,
            "convertible": "qty",
        },
        {
            "label": _("Balance Qty"),
            "fieldname": "qty_after_transaction",
            "fieldtype": "Float",
            "width": 100,
            "convertible": "qty",
        },
        {
            "label": _("Voucher #"),
            "fieldname": "voucher_no",
            "fieldtype": "Dynamic Link",
            "options": "voucher_type",
            "width": 200,
        },
    ]

    return columns, data

File: dox_development/dox_development/report/item_card_quantity_report/item_card_quantity_report.js
// Copyright (c) 2023, rakesh and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Item Card Quantity Report"] = {
	"filters": [
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"width": "80",
			"options": "Company",
			"default": frappe.defaults.get_default("company"),
			"hidden":true
		},
		{
			"fieldname": "item_code",
			"label": __("Item"),
			"fieldtype": "Link",
			"width": "80",
			"options": "Item",
			"reqd": 1,
			"width":"20px",
			"get_query": function() {
				return {
					query: "erpnext.controllers.queries.item_query",
				};
			},
		},
		{
			"fieldname":"warehouse",
			"label": __("Warehouse"),
			"fieldtype": "Link",
			"options": "Warehouse",
			"reqd": 1,
			"get_query": function() {
				const company = frappe.query_report.get_filter_value('company');
				return {
					filters: { 'company': company }
				}
			}
		},
		{
			"fieldname":"from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1,
			"default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
		},
		{
			"fieldname":"to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1,
			"default": frappe.datetime.get_today()
		},
		{
			"fieldname":"batch_no",
			"label": __("Batch No"),
			"fieldtype": "Link",
			"options": "Batch",
		},
		{
			"fieldname": "variant_of",
			"label": __("Variant"),
			"fieldtype": "Link",
			"options": "Item",
			"get_query": function() {
				return {
					filters: {
						'has_variants': 1
					}
				}
			}
		}
	],
	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname == "out_qty" && data && data.out_qty < 0) {
			value = "<span style='color:red'>" + value + "</span>";
		}
		else if (column.fieldname == "in_qty" && data && data.in_qty > 0) {
			value = "<span style='color:green'>" + value + "</span>";
		}

		return value;
	},
};
